### PR TITLE
feat(//core/converters): Add conversion support for torch.narrow()

### PR DESCRIPTION
**Pull request by peri044**
# Description

Adding conversion support for operator `torch.narrow(input, axis, start, length)`
The variable `start` can be a scalar or tensor. 
The test case when `start` is a tensor hasn't been implemented in the PR as the IR for this testcase wasn't being parsed by torch jit parser (due to start being prim::constant(value={2} for example). However, using trtorchexec, I was able to test that case as well and it works.

Fixes [#151](https://github.com/NVIDIA/TRTorch/issues/151)

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by narendasan** [comment]
Can you explain more about the missing test case issue?

**Comment by peri044** [comment]
For the following graph
```
class Narrow(torch.nn.Module):
    def __init__(self):
        super(Narrow, self).__init__()
        
    def forward(self, x):
        return torch.narrow(x, 1, torch.tensor(2), 2)
```
The JIT representation obtained from print logs (of trtorchexec after graph lowering) is 
```
graph(%x.1 : Tensor):
              %2 : Long() = prim::Constant[value={2}]()
              %3 : int = prim::Constant[value=1]() # narrow.py:13:31
              %4 : int = prim::Constant[value=2]() # narrow.py:13:47
              %5 : Tensor = aten::narrow(%x.1, %3, %2, %4) # narrow.py:13:15
              return (%5)
```
The error returned by `torch::jit::parseIR(graph, &*g)` in the test case is as follows
```
C++ exception with description "
Could not parse literal{:

      graph(%x.1 : Tensor):
              %2 : Long() = prim::Constant[value={2}]()
                                                 ~ <--- HERE
```
Looks like when there is a constant tensor ({2}) instead of scalar, this issue occurs.

**Comment by narendasan** [comment]
Is long a tensor here?

**Comment by narendasan** [comment]
Pass the tensor in as an input to the graph instead of an inline constant

**Comment by peri044** [comment]
I tried that but I faced a different error since args[2] (start) won't be an IValue, but an ITensor. So args[2].ITensor() (`nvinfer1::ITensor` instance) needs to be converted as a scalar integer. [https://github.com/NVIDIA/TRTorch/pull/188/files#diff-9d97a3640ddc39b7d6f7cf3e769dfa4dR91](https://github.com/NVIDIA/TRTorch/pull/188/files#diff-9d97a3640ddc39b7d6f7cf3e769dfa4dR91).
The startIdx should be a scalar integer to calculate the indices that need to be passed through gather layer. 
If we use args[2].ITensor(), we need to change the converter code a  bit, but I'm not sure if this would be general enough for a common usecase.

**Comment by narendasan** [comment]
Could we use ITensororFreeze? Seems like we need it as a constant anyway or can we not do the broadcasting in TensorRT?

**Comment by peri044** [comment]
`ITensorOrFreeze` returns a constant ITensor as output. What I wanted is an integer value `start` that can be passed through `torch::arange` to generate a torch tensor starting from `start` to `start + length`. This torch tensor (indices) is then converter as a constant weight tensor (nvinfer::ITensor) to tensorrt using converter::Weights.

**Comment by narendasan** [comment]
Is the Tensor guaranteed to be a 0d tensor? Also if you look at the 2nd concat test you should find a way to pass the static tensor as a argument that should not be a ITensor

**Comment by peri044** [comment]
As per our discussion, here is an [issue to track](https://github.com/NVIDIA/TRTorch/issues/189) the missing test case.

File: core/conversion/converters/impl/select.cpp [hunk 1]
```diff
@@ -46,6 +46,78 @@ auto select_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns()
...
+            return true;
+        }
+    }).pattern({
+        "aten::narrow.Tensor(Tensor(a) self, int dim, Tensor start, int length) -> Tensor(a)",
+        [](ConversionCtx* ctx, const torch::jit::Node* n, args& args) -> bool {
+            auto in = args[0].ITensor();
+            auto axis  = args[1].unwrapToInt();
+            torch::Tensor start = args[2].IValue()->toTensor().to(torch::kI32);
+            // TODO: Is there a better way to get data from 0-dim tensor ?
+            int startIdx = static_cast<int*>(start.data_ptr())[0];
```
[the remark this answers is not in the document]
Comment (peri044) [reply core/conversion/converters/impl/select.cpp]: Cannot index a 0-dim tensor here, but I replaced it with an alternative as per torch recommendation.

[the remark this answers is not in the document]
Comment (peri044) [reply core/conversion/converters/impl/select.cpp]: Done !!

File: core/conversion/converters/impl/select.cpp [hunk 2]
```diff
@@ -46,6 +46,78 @@ auto select_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns()
 
             LOG_DEBUG("Output tensor shape: " << out->getDimensions());
 
+            return true;
+        }
+    }).pattern({
+        "aten::narrow(Tensor(a) self, int dim, int start, int length) -> Tensor(a)",
```
[the remark this answers is not in the document]
Comment (peri044) [reply core/conversion/converters/impl/select.cpp:52]: Yeah for the scalar case of `start`, there is no separate specialization. It's just `aten::narrow` in the IR.

**Comment by narendasan** [comment]
@peri044 is this good to merge, minus the test that is missing?

**Comment by peri044** [comment]
Yes. This is good to merge !! The [issue ](https://github.com/NVIDIA/TRTorch/issues/189)tracks and details about the missing test case

**Review by narendasan — APPROVED** [review]
